src/arch/MovieTexture/MovieTexture_FFMpeg.cpp

```diff
@@ -164,93 +166,128 @@ float MovieDecoder_FFMpeg::GetTimestamp() const
 	}
 
 	// In a logical situation, this means that display is outpacing decoding.
-	if (display_frame_num_ >= static_cast<int>(frame_buffer_.size())) {
+	if (display_frame_num_ >= static_cast<int>(packet_buffer_.size())) {
+		return 0;
+	}
+
+	PacketHolder* packet = packet_buffer_[display_frame_num_].get();
+
+	// Sanity check.
+	if (packet == nullptr) {
 		return 0;
 	}
-	return frame_buffer_[display_frame_num_]->frame_timestamp;
+
+	std::lock_guard<std::mutex> lock(packet->lock);
+	return packet->frame_timestamp;
 }
 
 bool MovieDecoder_FFMpeg::IsCurrentFrameReady() {
-	// We're displaying faster than decoding. Do not even try to display the frame.
-	if (display_frame_num_ >= static_cast<int>(frame_buffer_.size())) {
+	// We're displaying faster than decoding. Do not even try to display the
+	// frame.
+	if (display_frame_num_ >= static_cast<int>(packet_buffer_.size())) {
 		return false;
 	}
-	// If the whole movie is decoded, then the frame is definitely ready.
-	if (end_of_file_) {
-		return true;
-	}
 
+	FrameHolder* frame = frame_buffer_[(display_frame_num_ + offset_) % frame_buffer_.size()].get();
+	// To make sure the frame doesn't change from under us.
+	std::lock_guard<std::mutex> lock(frame->lock);
 
-	std::lock_guard<std::mutex> lock(frame_buffer_[display_frame_num_]->lock);
-	if (frame_buffer_[display_frame_num_]->skip) {
-		LOG->Info("Frame %i not decoded, skipping...", display_frame_num_);
-		return true;
+	// In terms of how the sliding window works, displayed and ready are
+	// opposites. If the frame hasn't been displayed, it's ready. If it has
+	// been displayed, then it hasn't been overwritten and reset yet.
+	if (frame->displayed) {
+		LOG->Info("Frame %i not decoded, total frames: %i", display_frame_num_, total_frames_);
 	}
-	if (!frame_buffer_[display_frame_num_]->decoded) {
-		LOG->Info("Frame %i not decoded and was not skipped, total frames: %i", display_frame_num_, total_frames_);
+	return !frame->displayed;
+}
+
+int MovieDecoder_FFMpeg::HandleNextPacket() {
+	// If the decoder hit the end of the file, then that means the packet
+	// buffer is complete.
+	if (!end_of_file_) {
+		// Add in a new FrameBuffer entry, and lock it immediately.
+		packet_buffer_.emplace_back(std::make_unique<PacketHolder>());
+		std::unique_lock<std::mutex> lock(packet_buffer_.back()->lock);
+		int status = SendPacketToBuffer();
+		if (status < 0) {
+			lock.unlock();
+			return status;
+		}
+
+		// If the decoded packet is the end of file.
+		if (end_of_file_) {
+			// Release the mutex.
+			lock.unlock();
+
+			packet_buffer_.pop_back(); // Don't display an EoF frame.
+			// If we had to approximate the number of frames, set the actual
+			// total number of frames. This is benign even if we did have an
+			// accurate frame count at the start.
+			total_frames_ = packet_buffer_.size();
```

Comment:
I've been wondering if a circular buffer might be best here. It would not come with the overhead of memory allocation and deallocation. I'm not sure if it's worth the trouble to implement, though.

Reply:
I'm not so sure the performance is improved by a circular buffer. Memory usage/allocation should be constant with the sliding window or a circular buffer, since in both cases we'd have `total_frames_` number of packets and `frame_buffer_.size()` number of frames. And iiuc we won't be allocating or deallocating any new memory while the video is playing back.

The reason we need to know the number of frames is to handle the end of movie playing. In the original code, the end of file signal from ffmpeg was the end of display, but since decoding is decoupled from display, we can't use it at the same time anymore. It gets especially complicated when factoring in video looping, and cases where the video ends much earlier than the song. We could potentially mark that spot in the buffer when we see an end of file, but that's probably functionally equivalent to marking the total frames here.

The main benefit of a circular buffer is more on ease of handling looping tbh, since offset isn't really needed anymore. It probably would make the code easier to follow as well, but that's also arguable either way. Let me know though if you think I'm off the mark on any of this ^